Updated the UI of the Main Activity

VarunT11 opened the pull request:
## Small Description of Changes Performed

- Updated the **Theme** of **Main Activity** to `Theme.MaterialComponents.Light.NoActionBar`.
- Added the Different **Icons** required during the new UI Update as **Vector Assets**.
- Added **Menu Items** of **Drawer Menu** and **Toolbar Menu** of the Main Activity.
- Updated the **Layout** of the **Main Activity** and added a **Landscape** Mode variation.
- Updated the `MainActivity.java` File with the new Layout and implemented changes according to the User State.
- Updated the `ServerSettingsFragment.java`, `LoginFragment.java`, and `ODKServicesPropertyUtils.java` Files for updating the `CURRENT_USER_STATE` Property when performing changes in the Current Authentication Flow.

linl33 commented:
Depends on https://github.com/odk-x/androidlibrary/pull/206

lakshyagupta21 reviewed (commented):
Reviewed mostly the coding style.
Will comment more if I find anything which needs to be fixed. But overall everything looks good.

@VarunT11 In case you've picked the image resources from anywhere the internet then please do mention them and add their link too. Its always good to give credits to someone else's work.

Congrats @VarunT11  for your 1st GSoC PR :-)

wbrunette commented:
runtests

wbrunette commented:
@VarunT11 I will wait to do my review after your fixes for @lakshyagupta21 comments. When you think it's fixed please "resolve conversation".

VarunT11 commented:
@wbrunette @lakshyagupta21 I have made the requested changes. Please review them!

lakshyagupta21 commented on services_app/src/main/java/org/opendatakit/services/MainActivity.java:
```diff
@@ -114,21 +147,181 @@ protected void onResume() {
+  private boolean onToolbarMenuItemClicked(MenuItem item){
     int id = item.getItemId();
+
     if (id == R.id.action_sync) {
```
Could've used `if elseif`

lakshyagupta21 commented on services_app/src/main/java/org/opendatakit/services/MainActivity.java:
```diff
@@ -143,38 +336,60 @@ public boolean onOptionsItemSelected(MenuItem item) {
+
+    drawerLayout.closeDrawer(GravityCompat.START);
+
+    if(item.getItemId()==R.id.drawer_resolve_conflict){
```
Use `switch case` whenever your if/ifelse exceeds 3

linl33 replied:
The `if/else if` is fine here. Resource IDs will be non-final in an upcoming version of AGP. There is a lint warning for this:
 

> Resource IDs will be non-final in Android Gradle Plugin version 5.0, avoid using them in switch case statements
> Avoid the usage of resource IDs where constant expressions are required.
>
> A future version of the Android Gradle Plugin will generate R classes with non-constant IDs in order to improve the performance of incremental compilation.
> To suppress this error, use the issue id "NonConstantResourceId" as explained in the Suppressing Warnings and Errors section.

lakshyagupta21 commented on services_app/src/main/java/org/opendatakit/services/MainActivity.java:
```diff
@@ -143,38 +336,60 @@ public boolean onOptionsItemSelected(MenuItem item) {
-    if (id == R.id.action_settings) {
-
+    if(item.getItemId()==R.id.drawer_settings){
       Intent intent = new Intent(this, AppPropertiesActivity.class);
       intent.putExtra(IntentConsts.INTENT_KEY_APP_NAME, getAppName());
       startActivityForResult(intent, SETTINGS_ACTIVITY_RESULT_CODE);
```
I can see there are lot of intent related stuff happening, what do you think about having one utility classes having all these methods over there and you just pass configuration(activity class) and let that handle all intents for your application.

lakshyagupta21 commented on services_app/src/main/java/org/opendatakit/services/utilities/ODKServicesPropertyUtils.java:
```diff
@@ -38,6 +38,9 @@ public static void clearActiveUser(PropertiesSingleton props) {
       properties.put(CommonToolProperties.KEY_USERNAME, "");
       properties.put(CommonToolProperties.KEY_PASSWORD, "");
 
+      properties.put(CommonToolProperties.KEY_IS_USER_AUTHENTICATED,"false");
```
Is it possible to keep it as actual boolean value instead of wrapping boolean value in a string?

VarunT11 replied:
No, the **Properties** Class only stores the values as a String, so we cannot keep it as an actual boolean value.
Reference - https://docs.oracle.com/javase/7/docs/api/java/util/Properties.html

lakshyagupta21 commented on services_app/src/main/res/layout-land/activity_main.xml:
```diff
@@ -0,0 +1,272 @@
+                    android:layout_width="0dp"
+                    android:layout_height="wrap_content"
+                    android:layout_marginStart="24dp"
+                    android:layout_marginTop="8dp"
+                    android:layout_marginEnd="8dp"
+                    android:layout_marginBottom="8dp"
```
Create dimens in dimen.xml, it will make this value consistent everywhere and errors can be avoided at later stage when we change this value (and forget to change at other places).

linl33 replied:
Likewise for the colors. They should be defined in a separate xml file.

lakshyagupta21 reviewed (changes requested):
Sorry I missed few styling related comments last time, I've commented them now, lets also get @wbrunette input on this whether we should ensure strict styling in the code or not.

VarunT11 commented:
## Small Description of Changes Performed
- Added Different **User States** as **Enums** in `UserStata.java`.
- Added **Themes** for different types of **Views**.
- Added the required colors and the dimens in `colors.xml` and `dimens.xml` File.
- Implemented `OnClickListeners` in `MainActivity.java` File.
- Handled the UI-Updates by implementing different **User States**.
- Added comments to each method added.

**PS:** @wbrunette @lakshyagupta21 @linl33 I have made the requested changes. Please review the commits and suggest any changes to be performed.

lakshyagupta21 commented on services_app/src/main/res/layout/activity_main.xml:
```diff
@@ -1,19 +1,232 @@
-<RelativeLayout xmlns:android="http://schemas.android.com/apk/res/android"
+<androidx.drawerlayout.widget.DrawerLayout xmlns:android="http://schemas.android.com/apk/res/android"
```
This looks great now.

lakshyagupta21 commented on services_app/src/main/java/org/opendatakit/services/sync/actions/fragments/LoginFragment.java:
```diff
@@ -166,6 +167,7 @@ private void setNewCredentials() {
       Map<String, String> properties = new HashMap<String, String>();
       properties.put(CommonToolProperties.KEY_AUTHENTICATION_TYPE,
           getString(R.string.credential_type_username_password));
+      properties.put(CommonToolProperties.KEY_CURRENT_USER_STATE, UserState.AUTHENTICATED_USER.toString());
```
Check what `UserState.AUTHENTICATED_USER.toString()` prints you.
https://stackoverflow.com/a/13291109 refer to this one for better understanding.

lakshyagupta21 commented on services_app/src/main/java/org/opendatakit/services/MainActivity.java:
```diff
@@ -19,203 +19,438 @@
+        tvUsername.setText(username);
+    }
+
+    /**
+     * Handling the Visibility of Menu Items in the Navigation Drawer
+     * It also sets the Text of Sign-In button with respect to the User State
+     *
+     * @param resolve_visible        : The Visibility of Resolve Conflicts Item
+     * @param switch_sign_in_visible : The Visibility of Switch Sign-In Method Item
+     * @param update_cred_visible    : The Visibility of Update User Credentials Item
+     */
```
+1

lakshyagupta21 reviewed (approved):
Looks good to me, there are few minor things but we can do it later once we have time.

linl33 commented:
@VarunT11 Services crashes in Android 4.4 with your changes. Tested on Google API Android 4.4 emulator.
Stacktrace
```
2021-06-27 14:36:06.120 3383-3383/? E/AndroidRuntime: FATAL EXCEPTION: main
    Process: org.opendatakit.services, PID: 3383
    java.lang.RuntimeException: Unable to start activity ComponentInfo{org.opendatakit.services/org.opendatakit.services.MainActivity}: android.view.InflateException: Binary XML file line #226: Error inflating class com.google.android.material.navigation.NavigationView
        at android.app.ActivityThread.performLaunchActivity(ActivityThread.java:2195)
        at android.app.ActivityThread.handleLaunchActivity(ActivityThread.java:2245)
        at android.app.ActivityThread.access$800(ActivityThread.java:135)
        at android.app.ActivityThread$H.handleMessage(ActivityThread.java:1196)
        at android.os.Handler.dispatchMessage(Handler.java:102)
        at android.os.Looper.loop(Looper.java:136)
        at android.app.ActivityThread.main(ActivityThread.java:5017)
        at java.lang.reflect.Method.invokeNative(Native Method)
        at java.lang.reflect.Method.invoke(Method.java:515)
        at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:779)
        at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:595)
        at dalvik.system.NativeStart.main(Native Method)
     Caused by: android.view.InflateException: Binary XML file line #226: Error inflating class com.google.android.material.navigation.NavigationView
        at android.view.LayoutInflater.createView(LayoutInflater.java:621)
        at android.view.LayoutInflater.createViewFromTag(LayoutInflater.java:697)
        at android.view.LayoutInflater.rInflate(LayoutInflater.java:756)
        at android.view.LayoutInflater.inflate(LayoutInflater.java:492)
        at android.view.LayoutInflater.inflate(LayoutInflater.java:397)
        at android.view.LayoutInflater.inflate(LayoutInflater.java:353)
        at androidx.appcompat.app.AppCompatDelegateImpl.setContentView(AppCompatDelegateImpl.java:696)
        at androidx.appcompat.app.AppCompatActivity.setContentView(AppCompatActivity.java:170)
        at org.opendatakit.services.MainActivity.onCreate(MainActivity.java:203)
        at android.app.Activity.performCreate(Activity.java:5231)
        at android.app.Instrumentation.callActivityOnCreate(Instrumentation.java:1087)
        at android.app.ActivityThread.performLaunchActivity(ActivityThread.java:2159)
        at android.app.ActivityThread.handleLaunchActivity(ActivityThread.java:2245) 
        at android.app.ActivityThread.access$800(ActivityThread.java:135) 
        at android.app.ActivityThread$H.handleMessage(ActivityThread.java:1196) 
        at android.os.Handler.dispatchMessage(Handler.java:102) 
        at android.os.Looper.loop(Looper.java:136) 
        at android.app.ActivityThread.main(ActivityThread.java:5017) 
        at java.lang.reflect.Method.invokeNative(Native Method) 
        at java.lang.reflect.Method.invoke(Method.java:515) 
        at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:779) 
        at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:595) 
        at dalvik.system.NativeStart.main(Native Method) 
     Caused by: java.lang.reflect.InvocationTargetException
        at java.lang.reflect.Constructor.constructNative(Native Method)
        at java.lang.reflect.Constructor.newInstance(Constructor.java:423)
        at android.view.LayoutInflater.createView(LayoutInflater.java:595)
        at android.view.LayoutInflater.createViewFromTag(LayoutInflater.java:697) 
        at android.view.LayoutInflater.rInflate(LayoutInflater.java:756) 
        at android.view.LayoutInflater.inflate(LayoutInflater.java:492) 
        at android.view.LayoutInflater.inflate(LayoutInflater.java:397) 
        at android.view.LayoutInflater.inflate(LayoutInflater.java:353) 
        at androidx.appcompat.app.AppCompatDelegateImpl.setContentView(AppCompatDelegateImpl.java:696) 
        at androidx.appcompat.app.AppCompatActivity.setContentView(AppCompatActivity.java:170) 
        at org.opendatakit.services.MainActivity.onCreate(MainActivity.java:203) 
        at android.app.Activity.performCreate(Activity.java:5231) 
        at android.app.Instrumentation.callActivityOnCreate(Instrumentation.java:1087) 
        at android.app.ActivityThread.performLaunchActivity(ActivityThread.java:2159) 
        at android.app.ActivityThread.handleLaunchActivity(ActivityThread.java:2245) 
        at android.app.ActivityThread.access$800(ActivityThread.java:135) 
        at android.app.ActivityThread$H.handleMessage(ActivityThread.java:1196) 
        at android.os.Handler.dispatchMessage(Handler.java:102) 
        at android.os.Looper.loop(Looper.java:136) 
        at android.app.ActivityThread.main(ActivityThread.java:5017) 
        at java.lang.reflect.Method.invokeNative(Native Method) 
        at java.lang.reflect.Method.invoke(Method.java:515) 
        at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:779) 
        at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:595) 
        at dalvik.system.NativeStart.main(Native Method) 
     Caused by: android.view.InflateException: Binary XML file line #24: Error inflating class com.google.android.material.button.MaterialButton
        at android.view.LayoutInflater.createView(LayoutInflater.java:621)
        at android.view.LayoutInflater.createViewFromTag(LayoutInflater.java:697)
        at android.view.LayoutInflater.rInflate(LayoutInflater.java:756)
        at android.view.LayoutInflater.rInflate(LayoutInflater.java:759)
        at android.view.LayoutInflater.rInflate(LayoutInflater.java:759)
        at android.view.LayoutInflater.inflate(LayoutInflater.java:492)
        at android.view.LayoutInflater.inflate(LayoutInflater.java:397)
        at com.google.android.material.internal.NavigationMenuPresenter.inflateHeaderView(NavigationMenuPresenter.java:222)
        at com.google.android.material.navigation.NavigationView.inflateHeaderView(NavigationView.java:387)
        at com.google.android.material.navigation.NavigationView.<init>(NavigationView.java:241)
        at com.google.android.material.navigation.NavigationView.<init>(NavigationView.java:122)
        at java.lang.reflect.Constructor.constructNative(Native Method) 
        at java.lang.reflect.Constructor.newInstance(Constructor.java:423) 
        at android.view.LayoutInflater.createView(LayoutInflater.java:595) 
        at android.view.LayoutInflater.createViewFromTag(LayoutInflater.java:697) 
        at android.view.LayoutInflater.rInflate(LayoutInflater.java:756) 
        at android.view.LayoutInflater.inflate(LayoutInflater.java:492) 
        at android.view.LayoutInflater.inflate(LayoutInflater.java:397) 
        at android.view.LayoutInflater.inflate(LayoutInflater.java:353) 
        at androidx.appcompat.app.AppCompatDelegateImpl.setContentView(AppCompatDelegateImpl.java:696) 
        at androidx.appcompat.app.AppCompatActivity.setContentView(AppCompatActivity.java:170) 
        at org.opendatakit.services.MainActivity.onCreate(MainActivity.java:203) 
        at android.app.Activity.performCreate(Activity.java:5231) 
        at android.app.Instrumentation.callActivityOnCreate(Instrumentation.java:1087) 
        at android.app.ActivityThread.performLaunchActivity(ActivityThread.java:2159) 
        at android.app.ActivityThread.handleLaunchActivity(ActivityThread.java:2245) 
        at android.app.ActivityThread.access$800(ActivityThread.java:135) 
        at android.app.ActivityThread$H.handleMessage(ActivityThread.java:1196) 
        at android.os.Handler.dispatchMessage(Handler.java:102) 
        at android.os.Looper.loop(Looper.java:136) 
        at android.app.ActivityThread.main(ActivityThread.java:5017) 
        at java.lang.reflect.Method.invokeNative(Native Method) 
        at java.lang.reflect.Method.invoke(Method.java:515) 
        at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:779) 
        at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:595) 
        at dalvik.system.NativeStart.main(Native Method) 
     Caused by: java.lang.reflect.InvocationTargetException
        at java.lang.reflect.Constructor.constructNative(Native Method)
        at java.lang.reflect.Constructor.newInstance(Constructor.java:423)
        at android.view.LayoutInflater.createView(LayoutInflater.java:595)
        at android.view.LayoutInflater.createViewFromTag(LayoutInflater.java:697) 
        at android.view.LayoutInflater.rInflate(LayoutInflater.java:756) 
        at android.view.LayoutInflater.rInflate(LayoutInflater.java:759) 
        at android.view.LayoutInflater.rInflate(LayoutInflater.java:759) 
        at android.view.LayoutInflater.inflate(LayoutInflater.java:492) 
        at android.view.LayoutInflater.inflate(LayoutInflater.java:397) 
        at com.google.android.material.internal.NavigationMenuPresenter.inflateHeaderView(NavigationMenuPresenter.java:222) 
        at com.google.android.material.navigation.NavigationView.inflateHeaderView(NavigationView.java:387) 
        at com.google.android.material.navigation.NavigationView.<init>(NavigationView.java:241) 
        at com.google.android.material.navigation.NavigationView.<init>(NavigationView.java:122) 
        at java.lang.reflect.Constructor.constructNative(Native Method) 
        at java.lang.reflect.Constructor.newInstance(Constructor.java:423) 
        at android.view.LayoutInflater.createView(LayoutInflater.java:595) 
        at android.view.LayoutInflater.createViewFromTag(LayoutInflater.java:697) 
        at android.view.LayoutInflater.rInflate(LayoutInflater.java:756) 
        at android.view.LayoutInflater.inflate(LayoutInflater.java:492) 
        at android.view.LayoutInflater.inflate(LayoutInflater.java:397) 
        at android.view.LayoutInflater.inflate(LayoutInflater.java:353) 
        at androidx.appcompat.app.AppCompatDelegateImpl.setContentView(AppCompatDelegateImpl.java:696) 
        at androidx.appcompat.app.AppCompatActivity.setContentView(AppCompatActivity.java:170) 
        at org.opendatakit.services.MainActivity.onCreate(MainActivity.java:203) 
        at android.app.Activity.performCreate(Activity.java:5231) 
        at android.app.Instrumentation.callActivityOnCreate(Instrumentation.java:1087) 
        at android.app.ActivityThread.performLaunchActivity(ActivityThread.java:2159) 
        at android.app.ActivityThread.handleLaunchActivity(ActivityThread.java:2245) 
        at android.app.ActivityThread.access$800(ActivityThread.java:135) 
        at android.app.ActivityThread$H.handleMessage(ActivityThread.java:1196) 
        at android.os.Handler.dispatchMessage(Handler.java:102) 
        at android.os.Looper.loop(Looper.java:136) 
        at android.app.ActivityThread.main(ActivityThread.java:5017) 
        at java.lang.reflect.Method.invokeNative(Native Method) 
        at java.lang.reflect.Method.invoke(Method.java:515) 
        at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:779) 
        at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:595) 
        at dalvik.system.NativeStart.main(Native Method) 
     Caused by: android.content.res.Resources$NotFoundException: File res/drawable/ic_arrow.xml from drawable resource ID #0x7f08006d
        at android.content.res.Resources.loadDrawable(Resources.java:2101)
        at android.content.res.TypedArray.getDrawable(TypedArray.java:602)
        at android.widget.TextView.<init>(TextView.java:826)
    	at android.widget.Button.<init>(Button.
```

VarunT11 commented:
## Reason For Crash
- Mobile Phones and Emulators whose Android Version is less than **5.0 (API Level 21)** do **not** support **Vector Drawables** directly.
- Exception was generated on assigning the `android:drawableEnd` attribute to `@drawable/ic_arrow` of `MaterialButton` in the `drawer_header.xml` file.

## Small Description of Changes Performed
- Created a new **layout** file for API Version greater than or equal to 21.
- Removed the `android:drawableEnd` attribute from the regular layout file and added it in the v-21 file.

**PS: ** @linl33 The Crash has been fixed. Please review the commit and suggest any changes required.
